### ISSUE TITLE
Allow the Ignore*ForCodeCoverage attributes to support interfaces, abstract classes, and base classes

### ISSUE THREAD

**Issue by simensen** [sebastianbergmann/phpunit#5511]
For my specific case, I have a `::test()` method on my Event classes. This `::test()` method returns a factory I can use to easily build variations of the Event for testing purposes. However, both the factory classes and the `::test()` method itself do not need to be counted for code coverage.

My desired results:

 * Do not include any of the Event factories in code coverage
 * Do not include any of the Event `::test()` methods in code coverage
 * Not be required to specify concrete implementations in concrete tests

----

Here is my code:

```php
/**
 * @template T of TestObject
 */
interface EventWithTestFactory
{
    /**
     * @phpstan-return TestObject<T>
     */
    public static function test(): object;
}
```

```php
/**
 * @extends TestObject<ConversationUnblocked>
 */
class ConversationUnblockedFactory extends TestObject
{
    public static function new(): self
    {
        return self::ofType(ConversationUnblocked::class);
    }

    /**
     * @throws \ReflectionException
     */
    public function partial(
        ?ConversationId $conversationId = null
    ): self {
        return $this
            ->fromPartial(func_get_args())
            ->withDefaults(self::defaultDefinition(...));
    }

    public static function defaultDefinition(): array
    {
        return [
            'conversationId' => ConversationId::generate(...),
        ];
    }

    public function withConversationId(ConversationId $conversationId): self
    {
        return $this->withDefaults(['conversationId' => $conversationId]);
    }
}
```

```php
/**
 * @implements EventWithTestFactory<ConversationUnblockedFactory>
 */
final readonly class ConversationUnblocked implements Event, EventWithTestFactory
{
    use EdgeAwareReflectionSerializing;

    public function __construct(
    ) {
    }

    public static function test(): ConversationUnblockedFactory
    {
        return ConversationUnblockedFactory::new();
    }
}
```

My tests extend a base test case for each group of Events, depending on domain:

```php
abstract class ConversationTestCase extends TestCase
{
    // ...
}
```

```php
final class UnblockConversationTest extends ConversationTestCase
{
    // ...
}
```

----

The only way I've been able to get this to work is to define concrete class names in the actual test object like this:

```php
#[IgnoreClassForCodeCoverage(ConversationUnblockedFactory::class)]
#[IgnoreMethodForCodeCoverage(ConversationUnblocked::class, 'test')]
final class UnblockConversationTest extends ConversationTestCase
{
    // ...
}
```

Since I have several domains and dozens of events, what I hoped to was leverage the testing base classes and the interfaces/base classes for the code I want to ignore in coverage so that I just need to do the following:

```php
#[IgnoreClassForCodeCoverage(TestObject::class)]
#[IgnoreMethodForCodeCoverage(EventWithTestFactory::class, 'test')]
abstract class ConversationTestCase extends TestCase
{
    // ...
}
```

----

TL;DR

 * It looks like the attributes are not inherited by the concrete test cases.
 * It looks like the class specified in the attributes has to match the class name exactly (and not simply be an instance or implementation of).

Is it feasible or doable to address either of these points? Is there *another* way for me to do this with the current version of PHPUnit that isn't deprecated?

Is there a way to globally configure classes or interfaces – or specific methods on these – that can be ignored by code coverage?

**Comment by simensen** [comment]
> Is there a way to globally configure classes or interfaces – or specific methods on these – that can be ignored by code coverage?

The older `CodeCoverageIgnore` attribute was easier to use for what I'm trying to accomplish, but I understand why we are deprecating it. That said, it would be nice to be able to set these things in one place rather than on a test-by-test basis.

For example, it isn't clear if I specify `#[IgnoreClassForCodeCoverage(ConversationUnblockedFactory::class)]` in one test case, and actually reference `ConversationUnblockedFactory` in another test cast, whether that means  `ConversationUnblockedFactory` will be excluded from code coverage if I run code coverage just on the second test where I don't have `#[IgnoreClassForCodeCoverage(ConversationUnblockedFactory::class)]`.

I did find a temporary workaround for ignoring all the factories. It seems to work, but I'm unsure if this approach has any unforeseen issues. :) It's not my favorite way to handle this, but it's better than the alternative.

```xml
<phpunit>
  <source>
    <include>
      <directory suffix=".php">./src/Acme/</directory>
    </include>
    <exclude>
      <directory suffix="Factory.php">./src/Acme/Model/Clients/Client/Events/Factories/</directory>
      <directory suffix="Factory.php">./src/Acme/Model/Conversations/Conversation/Events/Factories/</directory>
      <directory suffix="Factory.php">./src/Acme/Model/HomebuyerOnboarding/Events/Factories/</directory>
      <directory suffix="Factory.php">./src/Acme/Model/Partners/Partner/Events/Factories/</directory>
      <directory suffix="Factory.php">./src/Acme/Model/Partners/PartnerInvitation/Events/Factories/</directory>
      <directory suffix="Factory.php">./src/Acme/Model/Partners/PartnerMemberInvitation/Events/Factories/</directory>
      <directory suffix="Factory.php">./src/Acme/Model/Programs/Program/Events/Factories/</directory>
    </exclude>
  </source>
</phpunit>
```

**Comment by sebastianbergmann** [comment]
> The older `CodeCoverageIgnore` attribute was easier to use for what I'm trying to accomplish, but I understand why we are deprecating it.

The `CodeCoverageIgnore` attribute will remain deprecated and will still be removed, but `IgnoreClassForCodeCoverage`, `IgnoreMethodForCodeCoverage`, and `IgnoreFunctionForCodeCoverage` attributes were a bad idea. Their deprecation is scheduled for PHPUnit 10.4 and is tracked [here](https://github.com/sebastianbergmann/phpunit/issues/5513).
